internal/config/config.go

```diff
@@ -475,3 +667,62 @@ func isIPv4(ip net.IP) bool {
 func isIPv6(ip net.IP) bool {
 	return ip.To16() != nil && ip.To4() == nil
 }
+
+// validatePeerAutodiscovery verifies that peer autodiscovery config is
+// specified for all required BGP params, or that default values are in place.
+//
+// Local ASN and peer ASN can be specified in annotations, labels or defaults.
+// Peer address can be specified in annotations or labels.
+func validatePeerAutodiscovery(p PeerAutodiscovery) error {
+	var localASNOK bool
+	var peerASNOK bool
+	var peerAddressOK bool
+
+	if d := p.Defaults; d != nil {
+		if d.MyASN != 0 {
+			localASNOK = true
+		}
+		if d.ASN != 0 {
+			peerASNOK = true
+		}
+		if d.Address != nil {
+			peerAddressOK = true
+		}
+	}
+
+	if a := p.FromAnnotations; a != nil {
+		if a.MyASN != "" {
+			localASNOK = true
+		}
+		if a.ASN != "" {
+			peerASNOK = true
+		}
+		if a.Addr != "" {
+			peerAddressOK = true
+		}
+	}
+
+	if l := p.FromLabels; l != nil {
+		if l.MyASN != "" {
+			localASNOK = true
+		}
+		if l.ASN != "" {
+			peerASNOK = true
+		}
+		if l.Addr != "" {
+			peerAddressOK = true
+		}
+	}
+
+	if !localASNOK {
```

Comment:
My initial comment from here still applies: https://github.com/metallb/metallb/pull/593/files?file-filters%5B%5D=.go&file-filters%5B%5D=.md#r439592048

Let me elaborate, but didn't have time to validate if what worries me actually happens in the code. So, I'll just share my concern and you tell me if it's an issue or not given how it is later used.

As I said in the linked discussion, these vars can be true but valid values for one can be set in annotations and for the other in labels.

For example, this can function can return true and all values in `p.fromAnnotations` can be valid except `p.FromAnnotations.MyASN` that is invalid. This will still return true because is valid in `p.FromLabels.MyASN`.

But IIUC, the [values from the annotation take precedence][link]. Therefore, if later for the peer `MyASN` values we use the in `FromAnnotations.MyASN` (seems legit, as this is a peer that passed validation and annotations takes precedence) this will fail because that value was not valid.

I haven't checked if, when using a parsedPeer, there is a logic (in this case) to use all values from Annotations, except `MyASN` from Labels as the `FromAnnotations.MyASN` is invalid. If there is such a logic, this seems fine. But if there isn't, this seems buggy.

The same applies for all the other fields.

IMHO, it seems semantically simpler if we validate that either `defaults + labels` are ok or `defaults+ annotations`. And my gut feeling is that code might be simpler for this too.

But, as I said, I haven't checked how this is used.

Is this a real issue? What do you think?

[link]:https://github.com/metallb/metallb/pull/593/files?file-filters%5B%5D=.go&file-filters%5B%5D=.md#diff-34d4dc28e4e4b8f9e871d6d68f62ef07R428-R429

Reply:
It would be really helpful if you pasted a sample config you suspect has a broken behavior :-) I'm trying to understand the specific case you are worried about. The text above confuses me. When I find it, I will add a unit test for it and fix the implementation if necessary.

Reply:
I think I get what you mean:

```yaml
peer-autodiscovery:
  from-annotations:
    peer-asn: example.com/peer-asn
    peer-address: example.com/peer-address
  from-labels:
    my-asn: example.com/my-asn
```

You are worried that with the config above we will try to use the `my-asn` specified in `from-annotations` rather than the one specified in `from-labels` (because annotations take precedence). This scenario doesn't happen because the precedence is **per parameter** which means the user is free to specify parameters in defaults, annotations or labels arbitrarily. For each parameter, the implementation of `parseNodePeer()` first assigns the value from `defaults` (if any), then overwrites it with the value from `from-labels` (if any) and finally overwrites it with the value from `from-annotations`. If the end result is that a required parameter (e.g. `my-asn`) is missing, `parseNodePeer()` errors out.

Here is a unit test which verifies this case:

https://github.com/kinvolk/metallb/blob/e4c388daf8d3f887356b23d479a29f56bc4a453b/speaker/bgp_controller_test.go#L1145-L1187

As you can see, in this test we specify the local ASN and the peer address in annotations and the peer ASN in labels, and the peer parsing succeeds.

IMO the existing behavior is [explained](https://github.com/kinvolk/metallb/blob/johananl/peer-autodiscovery/website/content/configuration/_index.md#peer-autodiscovery) clearly in the docs:

>This means that if the same BGP configuration parameter is specified using both an annotation and a label, the value from the annotation is used. A default value is used only if no relevant annotation or label is specified.

Reply:
@rata and I have discussed it online. The feedback above didn't take into account that `validatePeerAutodiscovery()` validates only the peer autodiscovery **mapping** and not the actual values provided in labels/annotations.